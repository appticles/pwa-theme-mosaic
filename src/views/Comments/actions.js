import config from 'react-global-configuration';

export const REQUEST_COMMENTS = 'REQUEST_COMMENTS';
export const RECEIVE_COMMENTS = 'RECEIVE_COMMENTS';

export const requestComments = () => ({
  type: REQUEST_COMMENTS,
});

export const receiveComments = comments => ({
  type: RECEIVE_COMMENTS,
  comments,
});

export const fetchComments = (params = {}) => (dispatch) => {
  dispatch(requestComments());

  let url;
  if (params) {
    url =
      config.get('export').comments +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receiveComments(json)))
    .catch(() => {
      dispatch(receiveComments([]));
    });
};

export const addComment = (params = {}) => {
  fetch(config.get('export').comments, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ params }),
  });
};
