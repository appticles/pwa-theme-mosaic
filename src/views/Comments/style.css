/* HEADER */
.ui.block.header {
  border: none;
  color: black;
  justify-content: center;
  font-size: 21px;
  font-weight: 100;
  position: fixed;
  margin: 0;
}

/* CLOSE ICON */
.ui.block.header a {
  position: absolute;
  right: 0;
}

/* COMMENTS */
.ui.comments {
  margin: 0;
}
.ui.comments .comment:first-of-type {
  margin: 4em 0;
}
.ui.comments .comment {
  height: unset;
  margin: 2em 0;
  padding: 0;
}

.ui.comments .comment > .avatar ~ .content {
  height: unset;
  text-align: start;
}

/* MODAL */
/* FIX */
.ui.comments + div {
  display: none;
}
/* override styling */
.dimmed.dimmable > .ui.animating.dimmer,
.dimmed.dimmable > .ui.visible.dimmer,
.ui.active.dimmer {
  opacity: 1;
}
.ui.modal.transition.visible.active:not(.share) {
  height: unset;
}

/* text */
.no-comments {
  padding-top: 20%;
}

/* Pull to refresh  */
.ui.comments .ui.block.header + div {
  height: unset;
}
