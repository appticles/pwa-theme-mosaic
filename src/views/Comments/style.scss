@import "../../utilities/variables";

.comments-container {
  /* HEADER */
  .ui.block.header.comments-header {
    border: none;
    text-align: center;
    margin: 0;
    color: $base-text-color;
    font-family: $base-font-family;
    /* CLOSE ICON */
    a {
      position: absolute;
      right: 0;
    }
  }
}
