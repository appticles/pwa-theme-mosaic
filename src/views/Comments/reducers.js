import { combineReducers } from 'redux';
import Immutable from 'seamless-immutable';
import _ from 'lodash';
import PropTypes from 'prop-types';

import { REQUEST_COMMENTS, RECEIVE_COMMENTS } from './actions';

export const commentPropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  post: PropTypes.number.isRequired,
  author_name: PropTypes.string.isRequired,
  content: PropTypes.shape({
    rendered: PropTypes.string.isRequired,
  }).isRequired,
});

export const INITIAL_STATE = Immutable({
  items: [],
  isFetching: 0,
});

const items = (state = INITIAL_STATE.items, action) => {
  switch (action.type) {
  case REQUEST_COMMENTS:
    return state;
  case RECEIVE_COMMENTS:
    if (Array.isArray(action.comments)) {
      return _.unionBy(state, action.comments, 'id');
    }
    return _.unionBy(state, [action.comments], 'id');
  default:
    return state;
  }
};
const isFetching = (state = INITIAL_STATE.isFetching, action) => {
  switch (action.type) {
  case REQUEST_COMMENTS:
    return state + 1;
  case RECEIVE_COMMENTS:
    return state - 1;
  default:
    return state;
  }
};

export const getComments = state => state.items;
export const getCommentsFetching = state => state.isFetching;

export default combineReducers({ items, isFetching });
