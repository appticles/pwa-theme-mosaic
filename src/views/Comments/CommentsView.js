import React from 'react';
import { Header, Comment, Icon, Container, Button, Modal } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import PropTypes from 'prop-types';

import { addComment } from './actions';
import { commentPropType } from './reducers';
import CommentForm from '../../components/Form/index';
import './style.css';

const CommentsView = (props) => {
  let path = {};
  if (_.isNil(props.match.params.categorySlug) || _.isNil(props.match.params.categoryId)) {
    path = '/post/' + props.match.params.postSlug + '/' + props.match.params.postId;
  } else {
    path =
      '/category/' +
      props.match.params.categorySlug +
      '/' +
      props.match.params.categoryId +
      '/post/' +
      props.match.params.postSlug +
      '/' +
      props.match.params.postId;
  }

  const submit = (values) => {
    const params = {
      author_name: values.name,
      email: values.email,
      content: {
        rendered: values.content,
      },
      date: new Date(),
      post: props.match.params.postId,
    };
    addComment(params);
    console.log(values);
  };
  console.log(props);
  return (
    <Comment.Group>
      <Header as="h3" block>
        Comments
        <Link to={path}>
          <Icon name="close" />
        </Link>
      </Header>
      {props.comments.length > 0 ? (
        props.comments.map(comment => (
          <Comment key={comment.id + comment.author}>
            <Comment.Avatar src={comment.author_avatar_urls[24]} />
            <Comment.Content>
              <Comment.Author as="a">{comment.author_name}</Comment.Author>
              <Comment.Metadata>
                <div>{comment.date}</div>
              </Comment.Metadata>
              <Comment.Text dangerouslySetInnerHTML={{ __html: comment.content.rendered }} />
              <Comment.Actions>
                <Comment.Action>Reply</Comment.Action>
              </Comment.Actions>
            </Comment.Content>
          </Comment>
        ))
      ) : (
        <Container className="no-comments">There are no comments for this article. Be the first one to comment about this!</Container>
      )}
      {props.match.params.comment_status === 'open' ? (
        <Modal trigger={<Button>Leave a comment</Button>} closeIcon>
          <Header icon="commenting" content="Leave a comment" />
          <Modal.Content>
            <CommentForm onSubmit={submit} />
          </Modal.Content>
        </Modal>
      ) : null}
    </Comment.Group>
  );
};

CommentsView.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      postSlug: PropTypes.string.isRequired,
      postId: PropTypes.string.isRequired,
      categorySlug: PropTypes.string,
      categoryId: PropTypes.string,
      comment_status: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  comments: PropTypes.arrayOf(commentPropType).isRequired,
};

export default CommentsView;
