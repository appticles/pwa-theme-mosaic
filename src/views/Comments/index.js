import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Loader, Header, Icon } from 'semantic-ui-react';
import _ from 'lodash';
import ReactPullToRefresh from 'react-pull-to-refresh';
import 'react-router-modal/css/react-router-modal.css';

import { fetchComments } from './actions';
import { getComments, getCommentsFetching, commentPropType } from './reducers';
import CommentsView from './components/CommentsView';

// translations
import { fetchTranslations } from '../../translations/actions';
import { getTranslations, getTranslationsFetching } from '../../translations/reducers';

import './style.css';

class Comments extends Component {
  constructor(props) {
    super(props);
    this.handleRefresh = this.handleRefresh.bind(this);
  }

  componentWillMount() {
    const { dispatch } = this.props;
    dispatch(fetchComments({ postId: this.props.match.params.postId, per_page: 100 }));
    dispatch(fetchTranslations);
  }

  handleRefresh(resolve, reject) {
    const { dispatch } = this.props;

    if (dispatch(fetchComments({ postId: this.props.match.params.postId, per_page: 100 }))) {
      resolve();
    } else {
      reject();
    }
  }

  render() {
    const comm = this.props.comments.filter(comment => comment.post === Number(this.props.match.params.postId));

    let path = '';
    if (!_.isNil(this.props.match.params.categorySlug) && !_.isNil(this.props.match.params.categoryId)) {
      path = '/category/' + this.props.match.params.categorySlug + '/' + this.props.match.params.categoryId;
    }

    path = path + '/post/' + this.props.match.params.postSlug + '/' + this.props.match.params.postId;

    return (
      <div className="comments-container">
        <Header as="h3" block className="comments-header">
          {this.props.translations.TEXTS && this.props.translations.TEXTS.COMMENTS}
          <Link to={path}>
            <Icon name="close" />
          </Link>
        </Header>
        {this.props.loading === 1 && this.props.loadingTranslations === 1 ? (
          <Loader active />
        ) : (
          <ReactPullToRefresh onRefresh={this.handleRefresh}>
            <CommentsView
              comments={comm}
              loading={this.props.loading}
              postId={this.props.match.params.postId}
              commentStatus={this.props.match.params.commentStatus}
              texts={this.props.translations}
            />
          </ReactPullToRefresh>
        )}
      </div>
    );
  }
}

Comments.propTypes = {
  dispatch: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      categoryId: PropTypes.string.isRequired,
      categorySlug: PropTypes.string.isRequired,
      postId: PropTypes.string.isRequired,
      postSlug: PropTypes.string.isRequired,
      commentStatus: PropTypes.oneOf(['open', 'closed', 'disabled']).isRequired,
    }).isRequired,
  }).isRequired,
  loading: PropTypes.number.isRequired,
  comments: PropTypes.arrayOf(commentPropType).isRequired,
  loadingTranslations: PropTypes.number.isRequired,
  translations: PropTypes.shape({
    TEXTS: PropTypes.shape({
      COMMENTS: PropTypes.string,
    }),
  }),
};

Comments.defaultProps = {
  translations: {
    TEXTS: {
      COMMENTS: 'Comments',
    },
  },
};

const mapStateToProps = state => ({
  comments: getComments(state.comments),
  loading: getCommentsFetching(state.comments),
  translations: getTranslations(state.translations),
  loadingTranslations: getTranslationsFetching(state.translations),
});

function mapDispatchToProps(dispatch) {
  return Object.assign({ dispatch }, bindActionCreators({ fetchComments, fetchTranslations }, dispatch));
}
export default connect(mapStateToProps, mapDispatchToProps)(Comments);
