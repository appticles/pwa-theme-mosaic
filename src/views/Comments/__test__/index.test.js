import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import mockStore from '../../../__mocks__/redux-mock-store';

import Comments from '../index';

jest.mock('../components/CommentsView', () => 'Comments view test');
jest.mock('../../../components/NotFound/index', () => 'Not found component test');

describe('Comments screen', () => {
  it('renders a message if the page does not exist', () => {
    const store = mockStore({
      comments: {
        items: [],
        isFetching: 0,
      },
      translations: {
        items: {},
        isFetching: 0,
      },
    });
    const match = {
      params: {
        categoryId: '42',
        categorySlug: 'category-slug',
        postId: '54',
        postSlug: 'post-slug',
      },
    };
    expect(
      renderer.create(
        <MemoryRouter>
          <Comments store={store} fetchComments={jest.fn()} match={match} />
        </MemoryRouter>,
      ),
    ).toMatchSnapshot();
  });
  it('renders comments list', () => {
    const store = mockStore({
      comments: {
        items: [
          {
            id: 67,
            post: 54,
            author_name: 'Comment Author',
            content: {
              rendered: 'Comment content',
            },
            date: 'd date',
            author_avatar_urls: {
              24: 'some-rurl',
            },
          },
        ],
        isFetching: 0,
      },
      translations: {
        items: {
          TEXTS: {
            COMMENTS: 'Comments',
          },
        },
        isFetching: 0,
      },
    });
    const match = {
      params: {
        categoryId: '42',
        categorySlug: 'category-slug',
        postId: '54',
        postSlug: 'post-slug',
      },
    };
    expect(
      renderer.create(
        <MemoryRouter>
          <Comments store={store} fetchComments={jest.fn()} match={match} />
        </MemoryRouter>,
      ),
    ).toMatchSnapshot();
  });
});
