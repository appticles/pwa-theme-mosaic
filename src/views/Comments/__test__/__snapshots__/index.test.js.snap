// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments screen renders a message if the page does not exist 1`] = `
<div
  className="ui comments"
>
  <h3
    className="ui block header"
  >
    <a
      href="/category/category-slug/42/post/post-slug/54"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="close icon"
      />
    </a>
  </h3>
  <div>
    <div
      className="ptr-element"
    >
      <span
        className="genericon genericon-next"
      />
      <div
        className="loading"
      >
        <span
          className="loading-ptr-1"
        />
        <span
          className="loading-ptr-2"
        />
        <span
          className="loading-ptr-3"
        />
      </div>
    </div>
    <div
      className="refresh-view"
    >
      <Comments view test
        comments={Array []}
        match={
          Object {
            "params": Object {
              "categoryId": "42",
              "categorySlug": "category-slug",
              "postId": "54",
              "postSlug": "post-slug",
            },
          }
        }
        texts={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Comments screen renders comments list 1`] = `
<div
  className="ui comments"
>
  <h3
    className="ui block header"
  >
    Comments
    <a
      href="/category/category-slug/42/post/post-slug/54"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="close icon"
      />
    </a>
  </h3>
  <div>
    <div
      className="ptr-element"
    >
      <span
        className="genericon genericon-next"
      />
      <div
        className="loading"
      >
        <span
          className="loading-ptr-1"
        />
        <span
          className="loading-ptr-2"
        />
        <span
          className="loading-ptr-3"
        />
      </div>
    </div>
    <div
      className="refresh-view"
    >
      <Comments view test
        comments={
          Array [
            Object {
              "author_avatar_urls": Object {
                "24": "some-rurl",
              },
              "author_name": "Comment Author",
              "content": Object {
                "rendered": "Comment content",
              },
              "date": "d date",
              "id": 67,
              "post": 54,
            },
          ]
        }
        match={
          Object {
            "params": Object {
              "categoryId": "42",
              "categorySlug": "category-slug",
              "postId": "54",
              "postSlug": "post-slug",
            },
          }
        }
        texts={
          Object {
            "TEXTS": Object {
              "COMMENTS": "Comments",
            },
          }
        }
      />
    </div>
  </div>
</div>
`;
